### PR TITLE
feat: Convert AppStateController to typescript

### LINKED ISSUE #25922: Convert AppStateController (implementation + tests) to TypeScript
### What is this about?

As a prerequisite for migrating `AppStateController` to BaseController v2, and to support the TypeScript migration effort, we want to convert `AppStateController` to TypeScript along with its tests.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

- All functions and methods are annotated with argument and return types.
- All properties and free variables are annotated with types.
- `#` is used for private properties and methods instead of `_`.
- The test file is converted to TypeScript along with the implementation file.

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

- https://github.com/MetaMask/metamask-extension/issues/23017

### PR DESCRIPTION
**Pull request by kanthesha**
## **Description**

As a prerequisite for migrating AppStateController to BaseController v2, and to support the TypeScript migration effort for the extension, we want to convert AppStateController to TypeScript along with its tests.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27572?quickstart=1)

## **Related issues**

Fixes: #25922 

## **Manual testing steps**
N/A

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/app-state-controller.ts [hunk 1]
```diff
@@ -0,0 +1,860 @@
...
+  // TODO: Remove this as soon as PreferencesController upgrade to BaseControllerV2 merges with develop
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  preferencesStore: any;
+  messenger: AppStateControllerMessenger;
+  // TODO: Replace `any` with type
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  extension: any;
+};
+
+const getInitState = (
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:154]: Do you think we can align this name to other controllers? i.e. `getDefaultAppStateControllerState`

File: app/scripts/controllers/app-state-controller.ts [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+  showBetaHeader: isBeta(),
+  showPermissionsTour: true,
+  showNetworkBanner: true,
+  showAccountBanner: true,
+  trezorModel: null,
+  onboardingDate: null,
+  newPrivacyPolicyToastClickedOrClosed: null,
+  newPrivacyPolicyToastShownDate: null,
+  hadAdvancedGasFeesSetPriorToMigration92_3: false,
+  ...initState,
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:178]: hmm, I'm aware that this was already here, but do you think we could move this to the end of the object?

Alternatively, do you know why this was placed here?

File: app/scripts/controllers/app-state-controller.ts (hunk 1 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state-controller.ts:154]: @mikesposito We split the work into two parts: first, converting to TypeScript, and then upgrading to BaseController v2. This refactoring is included as part of this ticket: https://github.com/orgs/MetaMask/projects/113?pane=issue&itemId=78942927

File: app/scripts/controllers/app-state-controller.ts (hunk 2 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state-controller.ts:178]: I believe it's intentional, as we want to ensure that these state properties are properly reset. For instance, this explains why `signatureSecurityAlertResponses` is placed at the bottom: https://github.com/MetaMask/metamask-extension/pull/22778/files

File: app/scripts/controllers/app-state-controller.ts [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+  signatureSecurityAlertResponses: {},
+  switchedNetworkDetails: null,
+  switchedNetworkNeverShowMessage: false,
+  currentExtensionPopupId: 0,
+});
+
+export default class AppStateController extends EventEmitter {
+  // TODO: Replace `any` with type
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private extension: any;
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:196]: add import for 
````
import { Browser } from 'webextension-polyfill';
```

```suggestion
     private extension: Browser
```

File: app/scripts/controllers/app-state-controller.ts [hunk 4]
```diff
@@ -0,0 +1,860 @@
...
+
+  private timer: NodeJS.Timeout | null;
+
+  private isUnlocked: () => boolean;
+
+  private waitingForUnlock: { resolve: () => void }[];
+
+  private messagingSystem: AppStateControllerMessenger;
+
+  private _approvalRequestId: string | null;
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:210]: can we use the `#` pattern for private variables... instead of `_` or it's planned for BaseController upgrade ?

File: app/scripts/controllers/app-state-controller.ts [hunk 5]
```diff
@@ -0,0 +1,860 @@
...
+  isUnlocked: () => boolean;
+  initState?: Partial<AppStateControllerState>;
+  onInactiveTimeout?: () => void;
+  // TODO: Remove this as soon as PreferencesController upgrade to BaseControllerV2 merges with develop
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  preferencesStore: any;
+  messenger: AppStateControllerMessenger;
+  // TODO: Replace `any` with type
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  extension: any;
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:151]: To fix the any on extension
```
import { Browser } from 'webextension-polyfill';
```

```suggestion
  private extension: Browser
```

File: app/scripts/controllers/app-state-controller.ts [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+};
+
+type AppStateControllerOptions = {
+  addUnlockListener: (callback: () => void) => void;
+  isUnlocked: () => boolean;
+  initState?: Partial<AppStateControllerState>;
+  onInactiveTimeout?: () => void;
+  // TODO: Remove this as soon as PreferencesController upgrade to BaseControllerV2 merges with develop
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  preferencesStore: any;
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:147]: I believe this can have the `ObservableStore<PreferencesControllerState>` type, regardless of the BaseControllerV2 upgrade

File: app/scripts/controllers/app-state-controller.ts (hunk 1 above)
Comment (mikesposito) [reply app/scripts/controllers/app-state-controller.ts:154]: Oh I see, makes sense!

File: app/scripts/controllers/app-state-controller.ts (hunk 2 above)
Comment (mikesposito) [reply app/scripts/controllers/app-state-controller.ts:178]: Hmm, I see. But then do you think that the `initState?: Partial<AppStateControllerState>` type might be misleading? Some of the keys passed as `initState` would be simply ignored

File: app/scripts/controllers/app-state-controller.ts (hunk 4 above)
Comment (mikesposito) [reply app/scripts/controllers/app-state-controller.ts:210]: I feel like this is more related to the typescript conversion than the BaseController upgrade. Though I agree that `#` variables should be preferred

File: app/scripts/controllers/app-state-controller.ts (hunk 2 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state-controller.ts:178]: You raise a valid point; we should exclude these properties to prevent any confusion I think.

File: app/scripts/controllers/app-state-controller.ts (hunk 5 above)
Comment (mikesposito) [reply app/scripts/controllers/app-state-controller.ts:151]: ^ @cryptodev-2s is `private` intentional in your suggestion?

File: app/scripts/controllers/app-state-controller.ts [hunk 7]
```diff
@@ -0,0 +1,860 @@
...
+    '0x539': true,
+  },
+  surveyLinkLastClickedOrClosed: null,
+  signatureSecurityAlertResponses: {},
+  switchedNetworkDetails: null,
+  switchedNetworkNeverShowMessage: false,
+  currentExtensionPopupId: 0,
+});
+
+export default class AppStateController extends EventEmitter {
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:193]: Curious: was this exported as default for a specific reason?

**Review by mikesposito — COMMENTED** [review]
We should be able to remove `app/script/app-state.d.ts` now

**Comment by mikesposito** [comment]
> This PR is blocked by [PreferencesController upgrade to BaseControllerV2](https://github.com/MetaMask/metamask-extension/pull/27398).

Curious: why is this dependent on PreferencesController upgrade?

File: app/scripts/controllers/app-state-controller.ts (hunk 5 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state-controller.ts:151]: @mikesposito , it's fixed, just a small typo. Initially, I was going to suggest placing this in the class variable, then moving it to options, and referencing the options in the class variable instead.

File: app/scripts/controllers/app-state-controller.ts [hunk 8]
```diff
@@ -0,0 +1,869 @@
...
+      this.emit(METAMASK_CONTROLLER_EVENTS.UPDATE_BADGE);
+    }
+
+    this._acceptApproval();
+  }
+
+  /**
+   * Sets the default home tab
+   *
+   * @param [defaultHomeActiveTabName] - the tab name
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:320]: Nit: Since we’ve defined the `defaultHomeActiveTabName` type, we can remove the square brackets `[]` from the JSDoc that marks the param as optional.
```suggestion
   * @param defaultHomeActiveTabName - the tab name
```

File: app/scripts/controllers/app-state-controller.ts [hunk 9]
```diff
@@ -0,0 +1,869 @@
...
+  setCurrentPopupId(currentPopupId: number): void {
+    this.store.updateState({
+      currentPopupId,
+    });
+  }
+
+  /**
+   * The function returns information about the last confirmation user interacted with
+   *
+   * @type {LastInteractedConfirmationInfo}: Information about the last confirmation user interacted with.
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:798]:
```suggestion
```

File: app/scripts/controllers/app-state-controller.ts [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+  getLastInteractedConfirmationInfo():
+    | LastInteractedConfirmationInfo
+    | undefined {
+    return this.store.getState().lastInteractedConfirmationInfo;
+  }
+
+  /**
+   * Update the information about the last confirmation user interacted with
+   *
+   * @type {LastInteractedConfirmationInfo} - information about transaction user last interacted with.
```
Comment (cryptodev-2s) [inline app/scripts/controllers/app-state-controller.ts:809]:
```suggestion
```

File: app/scripts/controllers/app-state-controller.ts [hunk 11]
```diff
@@ -0,0 +1,867 @@
...
+
+const controllerName = 'AppStateController';
+
+/**
+ * Returns the state of the {@link AppStateController}.
+ */
+export type AppStateControllerGetStateAction = {
+  type: 'AppStateController:getState';
+  handler: () => AppStateControllerState;
+};
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:87]: I don't think there is an handler registered for this event since this controller doesn't extend BaseController yet

File: app/scripts/controllers/app-state-controller.ts [hunk 12]
```diff
@@ -0,0 +1,867 @@
...
+  | AcceptRequest
+  | PreferencesControllerGetStateAction;
+
+/**
+ * Event emitted when the state of the {@link AppStateController} changes.
+ */
+export type AppStateControllerStateChangeEvent = {
+  type: 'AppStateController:stateChange';
+  payload: [AppStateControllerState, []];
+};
```
Comment (mikesposito) [inline app/scripts/controllers/app-state-controller.ts:108]: My understanding is that this event is not emitted through the messenger, but from EventEmitter - is that correct?

File: app/scripts/controllers/app-state-controller.ts (hunk 11 above)
Comment (cryptodev-2s) [reply app/scripts/controllers/app-state-controller.ts:87]: Actually @mikesposito is right, either remove the get state action or register an action handler at the controller level, but as this is not currently in use we can just remove it and can be easily added in BaseController upgrade. and event registration will be done automatically.

**Review by cryptodev-2s — DISMISSED** [review]

**Review by cryptodev-2s — DISMISSED** [review]

File: app/scripts/controllers/mmi-controller.ts [hunk 13]
```diff
@@ -70,8 +69,6 @@ export default class MMIController extends EventEmitter {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   public keyringController: any;
 
-  public preferencesController: PreferencesController;
```
Comment (mikesposito) [inline app/scripts/controllers/mmi-controller.ts:73]: Is this change related to `AppStateController`?
I think that this PR joins the `AppStateController` typescript conversion, an initial migration to BaseControllerV2, and also some post-merge work derived from PreferencesController

If that's correct, perhaps splitting these changes in multiple specific PR would make it easier to go through all the changes

**Review by cryptodev-2s — DISMISSED** [review]

**Review by cryptodev-2s — APPROVED** [review]

File: development/ts-migration-dashboard/files-to-convert.json [hunk 14]
```diff
@@ -4,7 +4,6 @@
   "app/scripts/constants/contracts.js",
   "app/scripts/constants/on-ramp.js",
   "app/scripts/contentscript.js",
-  "app/scripts/controllers/app-state.js",
```
Comment (DDDDDanica) [inline development/ts-migration-dashboard/files-to-convert.json:7]: 💯

**Review by DDDDDanica — APPROVED** [review]
